File: Applied Course/4.Problem Solving/Leetcode Callenge/3.Best Time to Buy and Sell Stock II.py
# https://leetcode.com/problems/best-time-to-buy-and-sell-stock-ii/
# Logic-1
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        def calculateProfit(prices,start):
            if start>=len(prices):
                return 0
            total_profit=0
            for i in range(start,len(prices)):
                max_profit=0
                for j in range(i+1,len(prices)):
                    if prices[i]<prices[j]:
                        profit=calculateProfit(prices,j+1)+prices[j]-prices[i]
                        if profit>max_profit:
                            max_profit=profit
                if max_profit>total_profit:
                    total_profit=max_profit
            return total_profit
        return calculateProfit(prices,0)
# Logic-2
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        i=0
        valley=prices[0]
        peak=prices[0]
        max_profit=0
        while i<len(prices)-1:
            while i<len(prices)-1 and prices[i]>=prices[i+1]:
                i+=1
            valley=prices[i]
            while i<len(prices)-1 and prices[i]<=prices[i+1]:
                i+=1
            peak=prices[i]
            max_profit+=peak-valley
        return max_profit
# Logic-3
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        max_profit=0
        for i in range(1,len(prices)):
            if prices[i]>prices[i-1]:
                max_profit+=prices[i]-prices[i-1]
        return max_profit

File: Applied Course/4.Problem Solving/Leetcode Callenge/3.Best Time to Buy and Sell Stock II.cpp
#include<bits/stdc++.h>
using namespace std;
// Logic-1
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int i=0,valley=prices[0],peak=prices[0],max_profit=0;
        while(i<prices.size()-1)
        {
            while(i<prices.size()-1 && prices[i]>=prices[i+1])
                i++;
            int valley=prices[i];
            while(i<prices.size()-1 && prices[i]<=prices[i+1])
                i++;
            int peak=prices[i];
            max_profit+=peak-valley;
        }
        return max_profit;
    }
};
// Logic-2
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int max_profit=0;
        for(int i=1;i<prices.size();i++)
            if(prices[i]>prices[i-1])
                max_profit+=prices[i]-prices[i-1];
        return max_profit;
    }
};
int main()
{
    // https://leetcode.com/problems/best-time-to-buy-and-sell-stock-ii/
    return 0;
}